tests/integration/ros3/test_ros3.py

```diff
@@ -0,0 +1,52 @@
+from pynwb import NWBHDF5IO
+from pynwb.testing import TestCase
+import requests
+
+
+def _search_dandi_assets(url, filepath):
+    response = requests.request("GET", url, headers={"Accept": "application/json"}).json()
+
+    for asset in response["results"]:
+        if filepath == asset["path"]:
+            return asset["asset_id"]
+
+    if response.get("next", None):
+        return _search_dandi_assets(response["next"], filepath)
+
+    return None
+
+
+def get_dandi_asset_id(dandiset_id, filepath):
+    url = f"https://api.dandiarchive.org/api/dandisets/{dandiset_id}/versions/draft/assets/"
+    asset_id = _search_dandi_assets(url, filepath)
+    if asset_id is None:
+        raise ValueError(f'path {filepath} not found in dandiset {dandiset_id}.')
+    return asset_id
+
+
+def get_dandi_s3_url(dandiset_id, filepath):
+    """Get the s3 location for any NWB file on DANDI"""
+
+    asset_id = get_dandi_asset_id(dandiset_id, filepath)
+    url = f"https://api.dandiarchive.org/api/dandisets/{dandiset_id}/versions/draft/assets/{asset_id}/download/"
+
+    s3_url = requests.request(url=url, method='head').url
+    if '?' in s3_url:
+        return s3_url[:s3_url.index('?')]
+    return s3_url
+
+
+class TestRos3Streaming(TestCase):
+    # requires h5py to be built with the ROS3 driver: conda install -c conda-forge h5py
+    # also requires requests package to be installed
+
+    def test_read(self):
+        # ephys, Svoboda Lab (450 kB)
+        dandiset_id = '000006'
+        filepath = 'sub-anm372795/sub-anm372795_ses-20170718.nwb'
+        s3_path = get_dandi_s3_url(dandiset_id, filepath)
```

Comment:
Rather than using an actual user asset, I think it would make sense for DANDI to have an S3 bucket with small test files. This is both to make this easier to control but also because running tests against actual data will throw off usage stats for DANDI.

Reply:
Having an S3 test bucket on DANDI would also allow testing against NWB files with different versions (as part of a nightly test).

Reply:
The test now points to `https://dandiarchive.s3.amazonaws.com/ros3test.nwb` assuming that @yarikoptic or someone else on the DANDI team can host that.

Reply:
Assuming this is an easy thing for the DANDI team to do, I'd like to hold the release until tomorrow so that we can get this PR in. @oruebel would that be all right?

Reply:
which file we should host?  I would prefer if you provide one so it is not encumbered by any possible fiasco with user wishes/license/suboptimal nwb version etc.  Just create one (or some) you would like to see representative of the features you would gain with ROS3 (not fetching "large" data arrays etc) and share and we will gladly host it/them

Reply:
BTW, might be worth testing that you/h5py can access it through a redirect. E.g. try on https://api.dandiarchive.org/api/assets/1c095f5f-d1e2-45db-b807-fdcfea08c6de/download/

as long as our API is stable, we (you) can actually even create a dedicated pynwb testing dandiset and  release it on dandiarchive and get a DOI and "guaranteed" availability in the future.  If we do not mess up -- it will be truly "guaranteed"! ;)

Reply:
> as long as our API is stable, we (you) can actually even create a dedicated pynwb testing dandiset and release it on dandiarchive and get a DOI and "guaranteed" availability in the future. If we do not mess up -- it will be truly "guaranteed"! ;)

would you prefer us to do that? it seems like it would clutter dandi, but it would be a nice test

Reply:
> BTW, might be worth testing that you/h5py can access it through a redirect.

FYI this does not work. 

```
Python 3.9.6 | packaged by conda-forge | (default, Jul 11 2021, 03:37:25) [MSC v.1916 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import h5py
>>> h5py.File('https://api.dandiarchive.org/api/assets/902af029-f411-40b5-81f1-67930d56ee05/download/', 'r', driver='ros3')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\Ryan\miniconda3\envs\dev\lib\site-packages\h5py\_hl\files.py", line 444, in __init__
    fid = make_fid(name, mode, userblock_size,
  File "C:\Users\Ryan\miniconda3\envs\dev\lib\site-packages\h5py\_hl\files.py", line 199, in make_fid
    fid = h5f.open(name, flags, fapl=fapl)
  File "h5py\_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py\_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py\h5f.pyx", line 100, in h5py.h5f.open
OSError: Unable to open file (could not convert found "Content-Length" response ("0"))
```

Reply:
In this PR, we add a utility function that uses a `requests.request` call to get around the redirect.

When constructing the asset URL, which of these is more stable / preferred (or does it not matter)?
f'https://api.dandiarchive.org/api/dandisets/{dandiset_id}/versions/draft/assets/{asset_id}/download/'
f'https://api.dandiarchive.org/api/assets/{asset_id}/download/'

Reply:
the hope is -- "not matter" ;-)

re redirect -- sorry, didn't know, and allowing for HEAD requests was resolved only recently: https://github.com/dandi/dandi-api/issues/380 ... so we might need to go all the way to HDF5 lib to see what they would say about redirects.  Sorry for leading you the rabbit hole ;)

Reply:
@rly - nothing is stable till you publish. but once published then you should expect some stability. 

regarding redirect, i'm 90% sure the hdf5 library doesn't support a redirect.

Reply:
That's a good point. Now that we have [published datasets](https://gui.dandiarchive.org/#/dandiset/search?search=published), maybe we should use one of those.

Reply:
filed https://github.com/HDFGroup/hdf5/issues/905

Reply:
After a dataset is published, can I still add files to it, or is it locked forever?

Reply:
you can think of it as software releases. you can add, but the published version contains only the assets it was published with.